File: kalitech-engine/src/main/resources/kalitech/engine/bootstrap/EngineProxy.js
"use strict";

// FILE: kalitech/engine/bootstrap/EngineProxy.js
// Author: KΛYLΛ

/**
 * ENGINE Proxy for Graal host EngineApi:
 *  - no globals
 *  - safe dynamic module storage inside ENGINE.modules (no collisions with host methods)
 *  - dynamic API access: ENGINE.api("input") and sugar ENGINE.input()
 *
 * Contract:
 *  - Host may export explicit methods (engine.input()), but it is NOT required.
 *  - Host must expose registry-based lookup via either:
 *      - engine.registry().api(id, type?) / engine.registry().api(id)
 *      - or engine.api(id) directly
 *
 * This proxy provides:
 *  - ENGINE.api(id): universal module lookup (host-first, then registry)
 *  - ENGINE.<id>(): sugar for ENGINE.api("<id>")
 */
function createEngineProxy(javaEngine) {
    if (!javaEngine) throw new Error("[bootstrap] createEngineProxy: javaEngine is required");

    const store = Object.create(null);
    const modules = Object.create(null);

    function hostHas(prop) {
        try {
            return prop in javaEngine;
        } catch (_) {
            return false;
        }
    }

    function isSafeKey(prop) {
        // Prevent weird symbol/meta keys from being treated as API ids.
        if (typeof prop !== "string") return false;
        if (prop.length === 0) return false;
        if (prop === "__proto__" || prop === "prototype" || prop === "constructor") return false;
        return true;
    }

    function getHostFn(name) {
        try {
            const v = javaEngine[name];
            return (typeof v === "function") ? v.bind(javaEngine) : null;
        } catch (_) {
            return null;
        }
    }

    // Universal API lookup. Works even if host doesn't provide engine.<id>() methods.
    function api(id) {
        id = String(id);

        // 1) If host has direct method engine[id](), prefer it (explicit host API).
        const direct = getHostFn(id);
        if (direct) return direct();

        // 2) If host has engine.api(id), use it (new host contract).
        const hostApi = getHostFn("api");
        if (hostApi) return hostApi(id);

        // 3) If host has registry(), use registry.api(id)
        const regFn = getHostFn("registry");
        if (regFn) {
            const reg = regFn();
            if (reg) {
                try {
                    // Accept both registry.api(id) and registry.api(id, type) (we only need id here).
                    const ra = reg.api;
                    if (typeof ra === "function") return ra.call(reg, id);
                } catch (_) {
                    // ignore and continue
                }
                try {
                    // fallback to get(id) returning entry with .api
                    const ge = reg.get;
                    if (typeof ge === "function") {
                        const e = ge.call(reg, id);
                        if (e && e.api) return e.api;
                    }
                } catch (_) {
                    // ignore
                }
            }
        }

        // 4) Not found
        return null;
    }

    // Sugar function factory for ENGINE.<id>()
    function apiGetterFn(key) {
        return function () {
            const out = api(key);
            if (!out) throw new Error("[ENGINE] api missing: '" + key + "'");
            return out;
        };
    }

    function setModule(key, value) {
        key = String(key);
        if (modules[key]) throw new Error("[ENGINE] duplicate module key '" + key + "'");
        modules[key] = value;

        // only create short alias if it doesn't collide with host
        // IMPORTANT: short alias is a FUNCTION getter, not the module object itself
        // to avoid conflicts with engine.<id>() style.
        if (!hostHas(key)) store[key] = value;
        return value;
    }

    function getModule(key) {
        return modules[String(key)] || null;
    }

    function hasModule(key) {
        return !!modules[String(key)];
    }

    // Expose helpers on ENGINE itself (stored, not host)
    store.modules = modules;
    store.setModule = setModule;
    store.getModule = getModule;
    store.hasModule = hasModule;
    store.__host__ = javaEngine;

    // Expose universal API method
    store.api = api;

    return new Proxy(javaEngine, {
        has(target, prop) {
            if (prop in store) return true;
            if (prop in modules) return true;
            return prop in target;
        },

        get(target, prop) {
            if (prop in store) return store[prop];
            if (prop in modules) return modules[prop];

            // Prefer host member
            let v;
            try {
                v = target[prop];
            } catch (_) {
                v = undefined;
            }

            // If host provides a function, bind it, BUT:
            // if it's a zero-arg api getter (engine.input()) returning null, fallback to registry-based api().
            if (typeof v === "function" && typeof prop === "string") {
                const bound = v.bind(target);

                // Wrap only for safe keys to avoid breaking random host functions.
                if (prop && prop !== "api" && prop !== "registry" && prop !== "getRegistry") {
                    return function () {
                        const out = bound.apply(target, arguments);
                        if (out !== null && out !== undefined) return out;

                        // fallback to universal API lookup
                        const a = store.api ? store.api(prop) : null;
                        if (a !== null && a !== undefined) return a;

                        return out; // keep original null
                    };
                }

                return bound;
            }

            if (v !== undefined) return v;

            // Dynamic sugar: ENGINE.<id>() => ENGINE.api("<id>")
            if (typeof prop === "string" && prop.length > 0) {
                return function () {
                    const out = store.api(prop);
                    if (!out) throw new Error("[ENGINE] api missing: '" + prop + "'");
                    return out;
                };
            }

            return undefined;
        },

        set(_target, prop, value) {
            // Never write to host.
            // If this key collides with host API, put it ONLY into modules.
            // If it doesn't collide, keep it as a normal dynamic member.
            if (hostHas(prop)) {
                modules[prop] = value;
                return true;
            }
            store[prop] = value;
            return true;
        },

        ownKeys(target) {
            const hostKeys = Reflect.ownKeys(target);
            const storeKeys = Reflect.ownKeys(store);
            const modKeys = Reflect.ownKeys(modules);

            const out = hostKeys.slice();
            for (let i = 0; i < storeKeys.length; i++) if (out.indexOf(storeKeys[i]) < 0) out.push(storeKeys[i]);
            for (let i = 0; i < modKeys.length; i++) if (out.indexOf(modKeys[i]) < 0) out.push(modKeys[i]);
            return out;
        },

        getOwnPropertyDescriptor(target, prop) {
            if (prop in store) {
                return {configurable: true, enumerable: true, writable: true, value: store[prop]};
            }
            if (prop in modules) {
                return {configurable: true, enumerable: true, writable: true, value: modules[prop]};
            }

            // Provide a virtual descriptor for dynamic api getters so tooling doesn't choke.
            if (!(prop in target) && isSafeKey(prop)) {
                return {configurable: true, enumerable: true, writable: false, value: apiGetterFn(prop)};
            }

            return Object.getOwnPropertyDescriptor(target, prop);
        }
    });
}

module.exports = {createEngineProxy};

File: kalitech-engine/src/main/resources/kalitech/engine/bootstrap/EngineProxy.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local Classes = luaRuntime.class
local Error = luaRuntime.Error
function createEngineProxy(self, javaEngine)
    if not javaEngine then
        error(
            Classes:construct(Error, "[bootstrap] createEngineProxy: javaEngine is required"),
            0
        )
    end
    local store = KObject:create(nil)
    local modules = KObject:create(nil)
    local function hostHas(self, prop)
        do
            local function lua_catch(_)
                return true, false
            end
            local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
                return true, javaEngine[prop] ~= nil
            end)
            if not lua_try then
                lua_hasReturned, lua_returnValue = lua_catch(lua_hasReturned)
            end
            if lua_hasReturned then
                return lua_returnValue
            end
        end
    end
    local function isSafeKey(self, prop)
        if KTypeOf(prop) ~= "string" then
            return false
        end
        if #prop == 0 then
            return false
        end
        if prop == "__proto__" or prop == "prototype" or prop == "constructor" then
            return false
        end
        return true
    end
    local function getHostFn(self, name)
        do
            local function lua_catch(_)
                return true, nil
            end
            local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
                local v = javaEngine[name]
                local lua_temp_0
                if KTypeOf(v) == "function" then
                    lua_temp_0 = KFunction:bind(v, javaEngine)
                else
                    lua_temp_0 = nil
                end
                return true, lua_temp_0
            end)
            if not lua_try then
                lua_hasReturned, lua_returnValue = lua_catch(lua_hasReturned)
            end
            if lua_hasReturned then
                return lua_returnValue
            end
        end
    end
    local function api(self, id)
        id = tostring(id)
        local direct = getHostFn(_G, id)
        if direct then
            return direct(_G)
        end
        local hostApi = getHostFn(_G, "api")
        if hostApi then
            return hostApi(_G, id)
        end
        local regFn = getHostFn(_G, "registry")
        if regFn then
            local reg = regFn(_G)
            if reg then
                do
                    local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
                        local ra = reg.api
                        if KTypeOf(ra) == "function" then
                            return true, KFunction:call(ra, reg, id)
                        end
                    end)
                    if lua_try and lua_hasReturned then
                        return lua_returnValue
                    end
                end
                do
                    local lua_try, lua_hasReturned, lua_returnValue = pcall(function()
                        local ge = reg.get
                        if KTypeOf(ge) == "function" then
                            local e = KFunction:call(ge, reg, id)
                            if e and e.api then
                                return true, e.api
                            end
                        end
                    end)
                    if lua_try and lua_hasReturned then
                        return lua_returnValue
                    end
                end
            end
        end
        return nil
    end
    local function apiGetterFn(self, key)
        return function(self)
            local out = api(_G, key)
            if not out then
                error(
                    Classes:construct(
                        Error,
                        ("[ENGINE] api missing: '" .. tostring(key)) .. "'"
                    ),
                    0
                )
            end
            return out
        end
    end
    local function setModule(self, key, value)
        key = tostring(key)
        if rawget(modules, key) ~= nil then
            error(
                Classes:construct(
                    Error,
                    ("[ENGINE] duplicate module key '" .. tostring(key)) .. "'"
                ),
                0
            )
        end
        modules[key] = value
        if not hostHas(_G, key) then
            store[key] = value
        end
        return value
    end
    local function getModule(self, key)
        return rawget(modules, tostring(key))
    end
    local function hasModule(self, key)
        return rawget(modules, tostring(key)) ~= nil
    end
    store.modules = modules
    store.setModule = setModule
    store.getModule = getModule
    store.hasModule = hasModule
    store.__host__ = javaEngine
    store.api = api
    return KProxy(
        _G,
        javaEngine,
        {
            has = function(self, target, prop)
                if store[prop] ~= nil then
                    return true
                end
                if modules[prop] ~= nil then
                    return true
                end
                return target[prop] ~= nil
            end,
            get = function(self, target, prop)
                if store[prop] ~= nil then
                    return store[prop]
                end
                if modules[prop] ~= nil then
                    return modules[prop]
                end
                local v
                do
                    local function lua_catch(_)
                        v = nil
                    end
                    local lua_try, lua_hasReturned = pcall(function()
                        v = target[prop]
                    end)
                    if not lua_try then
                        lua_catch(lua_hasReturned)
                    end
                end
                if KTypeOf(v) == "function" and KTypeOf(prop) == "string" then
                    local bound = KFunction:bind(v, target)
                    if prop and prop ~= "api" and prop ~= "registry" and prop ~= "getRegistry" then
                        return function(self, ...)
                            local args = {...}
                            local out = KFunction:apply(v, target, args)
                            if out ~= nil and out ~= nil then
                                return out
                            end
                            local lua_store_api_1
                            if store.api then
                                lua_store_api_1 = store:api(prop)
                            else
                                lua_store_api_1 = nil
                            end
                            local a = lua_store_api_1
                            if a ~= nil and a ~= nil then
                                return a
                            end
                            return out
                        end
                    end
                    return bound
                end
                if v ~= nil then
                    return v
                end
                if KTypeOf(prop) == "string" and #prop > 0 then
                    return function(self)
                        local out = store:api(prop)
                        if not out then
                            error(
                                Classes:construct(Error, ("[ENGINE] api missing: '" .. prop) .. "'"),
                                0
                            )
                        end
                        return out
                    end
                end
                return nil
            end,
            set = function(self, _target, prop, value)
                if hostHas(_G, prop) then
                    modules[prop] = value
                    return true
                end
                store[prop] = value
                return true
            end,
            ownKeys = function(self, target)
                local hostKeys = KObject:keys(target)
                local storeKeys = KObject:keys(store)
                local modKeys = KObject:keys(modules)
                local out = KArrayOps.slice(hostKeys)
                do
                    local i = 0
                    while i < KLength(storeKeys) do
                        if KArrayOps.indexOf(out, KIndex(storeKeys, i)) < 0 then
                            KArrayOps.push(out, KIndex(storeKeys, i))
                        end
                        i = i + 1
                    end
                end
                do
                    local i = 0
                    while i < KLength(modKeys) do
                        if KArrayOps.indexOf(out, KIndex(modKeys, i)) < 0 then
                            KArrayOps.push(out, KIndex(modKeys, i))
                        end
                        i = i + 1
                    end
                end
                return out
            end,
            getOwnPropertyDescriptor = function(self, target, prop)
                if store[prop] ~= nil then
                    return {configurable = true, enumerable = true, writable = true, value = store[prop]}
                end
                if modules[prop] ~= nil then
                    return {configurable = true, enumerable = true, writable = true, value = modules[prop]}
                end
                if not (target[prop] ~= nil) and isSafeKey(_G, prop) then
                    return {
                        configurable = true,
                        enumerable = true,
                        writable = false,
                        value = apiGetterFn(_G, prop)
                    }
                end
                return KObject:getOwnPropertyDescriptor(target, prop)
            end
        }
    )
end
local BootstrapEngineProxyApi = Classes:create()
BootstrapEngineProxyApi.name = "BootstrapEngineProxyApi"
BootstrapEngineProxyApi.prototype.createEngineProxy = createEngineProxy
return Classes:construct(BootstrapEngineProxyApi)
